internal/provider/resource_gitlab_project_environment.go

```diff
@@ -0,0 +1,175 @@
+				Type:         schema.TypeString,
+				Optional:     true,
+				ValidateFunc: validation.StringIsNotEmpty,
+			},
+			"state": {
+				Description: "State the environment is in. Accepted values: available or stopped.",
+				Type:        schema.TypeString,
+				Computed:    true,
+			},
+		},
+	}
+})
+
+func resourceGitlabProjectEnvironmentCreate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	name := d.Get("name").(string)
+	options := gitlab.CreateEnvironmentOptions{
+		Name: &name,
+	}
+	if externalURL, ok := d.GetOk("external_url"); ok {
+		options.ExternalURL = gitlab.String(externalURL.(string))
+	}
+
+	project := d.Get("project").(string)
+
+	log.Printf("[DEBUG] Project %s create gitlab environment %q", project, *options.Name)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.CreateEnvironment(project, &options)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			return diag.Errorf("feature Environments is not available")
+		}
+		return diag.FromErr(err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+
+	return resourceGitlabProjectEnvironmentRead(ctx, d, meta)
+}
+
+func resourceGitlabProjectEnvironmentRead(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] read gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+
+	if err == nil {
+		fmt.Printf("%T \n %v", environmentIDInt, environmentIDInt)
+	}
+
+	log.Printf("[DEBUG] Project %s read gitlab environment %d", project, environmentIDInt)
+
+	client := meta.(*gitlab.Client)
+
+	environment, resp, err := client.Environments.GetEnvironment(project, environmentIDInt)
+	if err != nil {
+		if resp != nil && resp.StatusCode == http.StatusNotFound {
+			log.Printf("[DEBUG] Project %s gitlab environment %q not found", project, environmentID)
+			d.SetId("")
+			return nil
+		}
+		return diag.Errorf("error getting gitlab project %q environment %q: %v", project, environmentID, err)
+	}
+
+	d.SetId(buildTwoPartID(&project, gitlab.String(fmt.Sprintf("%v", environment.ID))))
+	d.Set("project", project)
+	d.Set("name", environment.Name)
+	d.Set("state", environment.State)
+
+	return nil
+}
+
+func resourceGitlabProjectEnvironmentUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	log.Printf("[DEBUG] update gitlab environment %s", d.Id())
+
+	project, environmentID, err := parseTwoPartID(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	environmentIDInt, err := strconv.Atoi(environmentID)
+	if err != nil {
+		return diag.Errorf("error converting environment ID to int: %v", err)
+	}
+
+	name := d.Get("name").(string)
+	options := gitlab.EditEnvironmentOptions{
+		Name: &name,
```

Comment:
I'd also only send the environment name when it changed. 

Btw. here you could use `gitlab.String()` :)

Reply:
Do you mean when the External URL changed or when the name of the environment changes?  Environment name changes are to be removed completely in the future. See https://docs.gitlab.com/ee/ci/environments/#rename-an-environment